checker/src/summaries.rs

```diff
@@ -240,26 +255,64 @@ impl<'a, 'tcx: 'a> PersistentSummaryCache<'a, 'tcx> {
                 }
             }
         };
-        let tcx = self.type_context;
         let db = &self.db;
+        let tcx = self.type_context;
+        let persistent_key = self
+            .key_cache
+            .entry(def_id)
+            .or_insert_with(|| utils::summary_key_str(tcx, def_id));
+        let argument_key_cache = &self.argument_key_cache;
         self.cache.entry(def_id).or_insert_with(|| {
-            let persistent_key = utils::summary_key_str(tcx, def_id);
-            Self::get_persistent_summary_for_db(db, &persistent_key)
+            let arg_types_key = argument_key_cache.get(&def_id);
+            Self::get_persistent_summary_using_arg_types_if_possible(
+                db,
+                &persistent_key,
+                arg_types_key,
+            )
         })
     }
 
+    /// Returns a summary from the persistent summary cache, preferentially using the concatenation
+    /// of persistent_key with arg_types_key as the cache key and falling back to just the
+    /// persistent_key if arg_types_key is None.
+    fn get_persistent_summary_using_arg_types_if_possible(
+        db: &Db,
+        persistent_key: &str,
+        arg_types_key: Option<&String>,
+    ) -> Summary {
+        if let Some(arg_types_key) = arg_types_key {
+            if !arg_types_key.is_empty() {
```

Comment:
FYI you can probably use slice patterns here (https://doc.rust-lang.org/edition-guide/rust-2018/slice-patterns.html).

Reply:
Probably, but it does not seem a clear improvement.